### PR TITLE
Enable nullable for FileAppender, RollingFileAppender, and unit tests

### PR DESCRIPTION
**Pull request by FreeAndNil**
For #118 
And fix null issues. Also includes removing reflection code across unit tests for type safety, removing ArrayList usage in favor or strongly typed lists, and a few opportunistic auto-properties and other modernizations.
New version of #119
@erikmav I've created a new branch and PR because I couldn't resolve the conflicts in the old one.

### CODE REVIEW

File: src/log4net/Layout/Pattern/PatternLayoutConverter.cs [hunk 1]
```diff
@@ -94,17 +94,12 @@ public virtual bool IgnoresException
     /// </summary>
     /// <param name="writer"><see cref="TextWriter" /> that will receive the formatted result.</param>
     /// <param name="state">The state object on which the pattern converter should be executed.</param>
-    protected internal override void Convert(TextWriter writer, object state)
+    public override void Convert(TextWriter writer, object state)
     {
-      LoggingEvent loggingEvent = state as LoggingEvent;
-      if (loggingEvent != null)
-      {
+      if (state is LoggingEvent loggingEvent)
```
Comment (fluffynuts) [inline src/log4net/Layout/Pattern/PatternLayoutConverter.cs:99]: braces please - if you're using Rider or ReSharper, you can alt-enter on the `if` to get options - when you find the one to apply braces style, look at the menu to the right and "apply in file" (i'm normally too cautious to apply on a larger scale, but this beats alt-entering everything in the file)

File: src/log4net/Util/PatternStringConverters/PropertyPatternConverter.cs [hunk 2]
```diff
@@ -57,35 +57,24 @@ internal sealed class PropertyPatternConverter : PatternConverter
...
+      if (ThreadContext.Properties.GetProperties(false) is PropertiesDictionary threadProperties)
         compositeProperties.Add(threadProperties);
-      }
 
-      // TODO: Add Repository Properties
       compositeProperties.Add(GlobalContext.Properties.GetReadOnlyProperties());
 
-      if (Option != null)
-      {
+      if (Option is not null)
```
Comment (fluffynuts) [inline src/log4net/Util/PatternStringConverters/PropertyPatternConverter.cs:72]: braces please

File: src/log4net/Util/PatternStringConverters/UserNamePatternConverter.cs [hunk 3]
```diff
@@ -43,16 +43,12 @@ internal sealed class UserNamePatternConverter : PatternConverter
...
     {
       try
       {
-        System.Security.Principal.WindowsIdentity windowsIdentity = null;
-        windowsIdentity = System.Security.Principal.WindowsIdentity.GetCurrent();
-        if (windowsIdentity != null && windowsIdentity.Name != null)
-        {
-          writer.Write(windowsIdentity.Name);
-        }
+        if (System.Security.Principal.WindowsIdentity.GetCurrent()?.Name is string name)
```
Comment (fluffynuts) [inline src/log4net/Util/PatternStringConverters/UserNamePatternConverter.cs:50]: braces please

File: src/log4net/Appender/EventLogAppender.cs [hunk 4]
```diff
@@ -411,32 +411,21 @@ protected override void Append(LoggingEvent loggingEvent)
...
-      object categoryPropertyObj = loggingEvent.LookupProperty("Category");
-      if (categoryPropertyObj != null)
+      if (loggingEvent.LookupProperty("Category") is object categoryPropertyObj)
       {
-        if (categoryPropertyObj is short)
-        {
-          category = (short)categoryPropertyObj;
-        }
+        if (categoryPropertyObj is short shortValue)
+          category = shortValue;
```
[the remark this answers is not in the document]
Comment (FreeAndNil) [reply src/log4net/Appender/EventLogAppender.cs:417]: check

File: src/log4net/Appender/RollingFileAppender.cs [hunk 5]
```diff
@@ -589,8 +583,8 @@ protected virtual void AdjustFileBeforeAppend()
       }
       finally
       {
-          // if rolling should be locked, release the lock
-          m_mutexForRolling?.ReleaseMutex();
+        // if rolling should be locked, release the lock
+        m_mutexForRolling?.ReleaseMutex();
```
[the remark this answers is not in the document]
Comment (FreeAndNil) [reply src/log4net/Appender/RollingFileAppender.cs:587]: I don't think disposing the mutex at this point is a good idea.
The mutex is a class member and will be used again in every AdjustFileBeforeAppend() later.
I've adjusted the finalizer instead.

File: src/log4net/Appender/RollingFileAppender.cs [hunk 6]
```diff
@@ -222,7 +220,7 @@ protected internal enum RollPoint
       TopOfMonth = 5
     }
 
-    #endregion Protected Enums
+    #endregion
```
[the remark this answers is not in the document]
Comment (FreeAndNil) [reply src/log4net/Appender/RollingFileAppender.cs:223]: I've removed all regions and moved the date classes to an own file.

**Review by fluffynuts — APPROVED** [review]

**Comment by FreeAndNil** [comment]
> @FreeAndNil I was expanding nullable and strongly typed data structures in the code yesterday. Would it make sense for me to send it along incrementally with new PRs against the 111- branch? Or are you moving fast there and want to minimize interruption?

@erikmav your work is very welcome. One remark: we need to have code coverage for the changed blocks. Most blocks are covered but some are not, so adding missing unit tests is needed.

**Pull request merged**